leveltranslator.go

```diff
@@ -0,0 +1,193 @@
+package pdk
+
+import (
+	"encoding/binary"
+	"hash/fnv"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"os"
+
+	"github.com/pkg/errors"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/opt"
+)
+
+// LevelTranslator is a Translator which stores the two way val/id mapping in
+// leveldb.
+type LevelTranslator struct {
+	fmu    sync.RWMutex
+	frames map[string]mapDBs
+}
+
+type mapDBs struct {
+	lock   ValueLocker
+	idMap  *leveldb.DB
+	valMap *leveldb.DB
+	curID  *uint64
+}
+
+type Errors []error
+
+func (errs Errors) Error() string {
+	errstrings := make([]string, len(errs))
+	for i, err := range errs {
+		errstrings[i] = err.Error()
+	}
+	return strings.Join(errstrings, "; ")
+}
+
+func (lt *LevelTranslator) Close() error {
+	errs := make(Errors, 0)
+	for f, dbs := range lt.frames {
+		err := dbs.idMap.Close()
+		if err != nil {
+			errs = append(errs, errors.Wrapf(err, "closing idMap for frame: %v", f))
+		}
+		err = dbs.valMap.Close()
+		if err != nil {
+			errs = append(errs, errors.Wrapf(err, "closing valMap for frame: %v", f))
+		}
+	}
+	if len(errs) > 0 {
+		return errs
+	}
+	return nil
+}
+
+func NewLevelTranslator(dirname string, frames ...string) (lt *LevelTranslator, err error) {
+	lt = &LevelTranslator{
+		frames: make(map[string]mapDBs),
+	}
+	err = os.MkdirAll(dirname, 0700)
+	if err != nil {
+		return nil, errors.Wrap(err, "making directory")
+	}
+	for _, frame := range frames {
+		var initialID uint64 = 0
+		mdbs := mapDBs{
+			curID: &initialID,
+			lock:  NewBucketVLock(),
+		}
+		mdbs.idMap, err = leveldb.OpenFile(dirname+"/"+frame+"-id", &opt.Options{})
+		if err != nil {
+			return nil, errors.Wrapf(err, "opening leveldb at %v", dirname+"/"+frame+"-id")
+		}
+		mdbs.valMap, err = leveldb.OpenFile(dirname+"/"+frame+"-val", &opt.Options{})
+		if err != nil {
+			return nil, errors.Wrapf(err, "opening leveldb at %v", dirname+"/"+frame+"-val")
+		}
+		lt.frames[frame] = mdbs
+	}
+	return lt, err
+}
+
+func (lt *LevelTranslator) Get(frame string, id uint64) (val interface{}) {
+	var dbs mapDBs
+	var ok bool
+	if dbs, ok = lt.frames[frame]; !ok {
+		panic(errors.Errorf("frame %v not found in level translator", frame))
+	}
+	idBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(idBytes, id)
+	data, err := dbs.idMap.Get(idBytes, nil)
+	if err != nil {
+		panic(err)
+	}
+	return data
+}
+
+func (lt *LevelTranslator) GetID(frame string, val interface{}) (id uint64, err error) {
+	var dbs mapDBs
+	var ok bool
+	if dbs, ok = lt.frames[frame]; !ok {
+		return 0, errors.Errorf("frame %v not found in level translator", frame)
+	}
+	var valBytes []byte
+	switch valt := val.(type) {
+	case []byte:
+		valBytes = valt
+	case string:
+		valBytes = []byte(valt)
+	default:
+		return 0, errors.Errorf("val needs to be of type []byte, but is type: %T, val: '%v'", val, val)
+	}
+	var data []byte
+
+	// data, err = dbs.valMap.Get(valBytes, &opt.ReadOptions{})
```

Comment:
do these comments need to stay here?

Reply:
I will uncomment this, but for initial import in the use case I was testing it was faster not to have it.